File: src/main/java/com/kishlaly/ta/utils/Context.java
package com.kishlaly.ta.utils;

import com.kishlaly.ta.analyze.testing.sl.StopLossFixedPrice;
import com.kishlaly.ta.analyze.testing.sl.StopLossStrategy;
import com.kishlaly.ta.analyze.testing.tp.TakeProfitFixedKeltnerTop;
import com.kishlaly.ta.analyze.testing.tp.TakeProfitStrategy;
import com.kishlaly.ta.model.SymbolsSource;
import com.kishlaly.ta.model.Timeframe;

import java.util.ArrayList;
import java.util.List;
import java.util.Set;

public class Context {

    public enum ApiSource {
        RAPID, ALPHAVANTAGE, MARKETSTACK
    }

    public static String outputFolder = "/Users/volodymyr/Investment/TA";
    public static ApiSource api = ApiSource.ALPHAVANTAGE;
    public static double parallelRequests = 10;
    public static double limitPerMinute = 75;
    public static Timeframe timeframe = Timeframe.DAY; // по умолчению для тестов
    public static Timeframe aggregationTimeframe = Timeframe.DAY; // на основе этого таймфрема агрегируются другие котировки
    public static SymbolsSource[] source = {SymbolsSource.SP500, SymbolsSource.NAGA, SymbolsSource.SCREENER_FILTERED};
    public static List<String> testOnly = new ArrayList<>();
    public static Set<String> symbols;
    public static int yearsToAnalyze = 5;

    // чтобы конвертировать в мою таймзону и сравнивать с графиками TradingView
    public static String myTimezone = "Europe/Berlin";

    // время работы NYSE и NASDAQ шесть с половиной часов или 390 минут
    // используется для дневных баров
    //public static int workingTime = 390;
    public static int workingTime = 1440; // 1440 минут в сутках

    // если меньше, то пропускать эту акцию
    public static int minimumWeekBarsCount = 50;
    public static int minimumDayBarsCount = 100;
    public static int minimumHourBarsCount = 300;

    // для логов
    public static Timeframe logTimeframe1;
    public static Timeframe logTimeframe2;
    public static Timeframe runGroups;

    // для тестирования на истоирческих данных
    public static boolean testMode;
    public static double lots = 100;
    public static StopLossStrategy stopLossStrategy = new StopLossFixedPrice(0.27);
    public static TakeProfitStrategy takeProfitStrategy = new TakeProfitFixedKeltnerTop(80);
    public static boolean massTesting;
    public static List<TakeProfitStrategy> takeProfitStrategies;

    // разное
    public static boolean trendCheckIncludeHistogram = true;
    public static boolean TRIM_DATA = true;

}

File: src/main/kotlin/com/kishlaly/ta/config/Context.kt
package com.kishlaly.ta.config

import com.kishlaly.ta.analyze.testing.sl.StopLossFixedPrice
import com.kishlaly.ta.analyze.testing.sl.StopLossStrategy
import com.kishlaly.ta.analyze.testing.tp.TakeProfitFixedKeltnerTop
import com.kishlaly.ta.analyze.testing.tp.TakeProfitStrategy
import com.kishlaly.ta.model.ApiSource
import com.kishlaly.ta.model.SymbolsSource
import com.kishlaly.ta.model.Timeframe
import org.ktorm.database.Database

class Context {

    companion object {
        var accountBalance = 10000.0 // maximum position size

        var tradeCommission = 1 // %

        var basicTimeframes = ThreadLocal.withInitial { arrayOf(arrayOf(Timeframe.WEEK, Timeframe.DAY)) }
        var outputFolder = "data"
        var api = ApiSource.ALPHAVANTAGE
        var parallelRequests = 10.0
        var limitPerMinute = 75.0
        var timeframe = ThreadLocal.withInitial { Timeframe.DAY }

        // main aggregation timeframe. Weekly quotes will be calculated from it
        var aggregationTimeframe = ThreadLocal.withInitial { Timeframe.DAY }
        var source = arrayOf(SymbolsSource.SP500)
        var testOnly: List<String> = ArrayList()
        var symbols = ThreadLocal.withInitial { setOf<String>() }
        var yearsToAnalyze = 5
        var lowPricesOnly = false
        var lowPriceLimit = 20

        // to convert to my timezone and compare with TradingView charts
        var myTimezone = "Europe/Berlin"

        var workingTime = 1440 // 1440 minutes in a day

        // If less, then skip this stock
        var minimumWeekBarsCount = 50
        var minimumDayBarsCount = 100
        var minimumHourBarsCount = 300

        // for logs
        var logTimeframe1: ThreadLocal<Timeframe> = ThreadLocal.withInitial { null }
        var logTimeframe2: ThreadLocal<Timeframe> = ThreadLocal.withInitial { null }
        var runGroups: Timeframe? = null
        var useDBLogging: Boolean = false
        var database: Database? = null

        // for testing on historical data
        var testMode = false
        var stopLossStrategy: ThreadLocal<StopLossStrategy> = ThreadLocal.withInitial { StopLossFixedPrice(0.27) }
        var takeProfitStrategy: ThreadLocal<TakeProfitStrategy> =
            ThreadLocal.withInitial { TakeProfitFixedKeltnerTop(80) }
        var massTesting = false
        var takeProfitStrategies: MutableList<TakeProfitStrategy> = mutableListOf()

        // misc
        var trendCheckIncludeHistogram = true
        var TRIM_DATA = true
        var trimToDate: String? = null // format: dd.mm.yyyy
        var fileSeparator = System.getProperty("file.separator")

        const val TESTS_FOLDER = "tests"
        const val SINGLE_TXT = "single.txt"
        const val MASS_TXT = "mass.txt"
        const val MIN_POSSIBLE_QUOTES = 11

    }

}